### PR TITLE
[SPARK-39910][SQL] Delegate path qualification to filesystem during DataSource file path globbing

### PR DESCRIPTION
**Pull request by tigrulya-exe**
### What changes were proposed in this pull request?

In current version `DataSource#checkAndGlobPathIfNecessary` qualifies paths via `Path#makeQualified` and `PartitioningAwareFileIndex` qualifies via `FileSystem#makeQualified`. Most `FileSystem` implementations simply delegate to `Path#makeQualified`, but others, like `HarFileSystem` contain fs-specific logic, that can produce different result. Such inconsistencies can lead to a situation, when spark can't find partitions of the source file, because qualified paths, built by `Path` and `FileSystem` are different. Therefore, for uniformity, the `FileSystem` path qualification should be used in `DataSource#checkAndGlobPathIfNecessary`.  

### Why are the changes needed?

Allow users to read files from hadoop archives (.har) using DataFrameReader API

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New tests were added in `DataSourceSuite` and `DataFrameReaderWriterSuite`

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala [hunk 1]
```diff
@@ -25,6 +27,7 @@ import org.apache.spark.sql.AnalysisException
 import org.apache.spark.sql.test.SharedSparkSession
 
 class DataSourceSuite extends SharedSparkSession with PrivateMethodTester {
+
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:30]: Please restore this line.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala [hunk 2]
```diff
@@ -156,6 +159,39 @@ class DataSourceSuite extends SharedSparkSession with PrivateMethodTester {
...
+      absoluteHarPaths.map(_.toString),
+      hadoopConf,
+      checkEmptyGlobPath = true,
+      checkFilesExist = true,
+      enableGlobbing = true
+    )
+
+    assert(
+      resultPaths.toSet === absoluteHarPaths.toSet
+    )
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:176]:
```suggestion
    assert(resultPaths.toSet === absoluteHarPaths.toSet)
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala [hunk 3]
```diff
@@ -156,6 +159,39 @@ class DataSourceSuite extends SharedSparkSession with PrivateMethodTester {
...
+      hadoopConf,
+      checkEmptyGlobPath = true,
+      checkFilesExist = true,
+      enableGlobbing = true
+    )
+
+    val expectedHarPaths = buildFullHarPaths(allRelativeHarPaths)
+    assert(
+      resultPaths.toSet === expectedHarPaths.toSet
+    )
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:193]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala [hunk 4]
```diff
@@ -197,11 +233,35 @@ object TestPaths {
       )
   )
 
+  val txtRelativeHarPath = new Path("/test.txt")
+  val csvRelativeHarPath = new Path("/test.csv")
+  val jsonRelativeHarPath = new Path("/test.json")
+  val parquetRelativeHarPath = new Path("/test.parquet")
+  val orcRelativeHarPath = new Path("/test.orc")
+  val globeRelativeHarPath = new Path("/test.*")
```
Comment (beliefer) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:241]: Do we really need to test all the file format?

Comment (tigrulya-exe) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:236]: They're needed to test globe path. I decided to reuse har from `DataFrameReaderWriterSuite` tests instead of creating new archive.

**Comment by beliefer** [comment]
@tigrulya-exe Please re-trigger GA tests.

**Comment by tigrulya-exe** [comment]
@beliefer I re-ran tests several times, but they failed either due to lack of resources or due to flaky ProtobufCatalystDataConversionSuite. I will rebase on #43493 after it will be merged

**Comment by beliefer** [comment]
Please fix the conflicts.

**Comment by tigrulya-exe** [comment]
@beliefer Hi! I fixed the conflicts and rebased on master

**Review by beliefer — COMMENTED** [review]
cc @cloud-fan

**Comment by tigrulya-exe** [comment]
@cloud-fan Hi! Could you take a look please?

**Comment by tigrulya-exe** [comment]
@cloud-fan Hi! Could you take a look please?

**Comment by tigrulya-exe** [comment]
@cloud-fan Hi! I've rebased on master and fixed conflicts. Could you please take a look?

**Comment by cloud-fan** [comment]
The fix is straightforward but the test is convoluted. How do you test `HarFileSystem`? I can't find any code setting the file system, but only the directory name contains `har`.

File: sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala [hunk 5]
```diff
@@ -1363,4 +1363,19 @@ class DataFrameReaderWriterSuite extends QueryTest with SharedSparkSession with
       }
     }
   }
+
+  test("SPARK-39910: read files from Hadoop archives") {
+    val fileSchema = new StructType().add("str", StringType)
+    val harPath = testFile("test-data/test-archive.har")
+      .replaceFirst("file:/", "har:/")
+
+    testRead(spark.read.textFile(s"$harPath/test.txt").toDF(), data, textSchema)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala:1372]: Since we only want to test path globbing, I think testing one file format is sufficient.

Comment (tigrulya-exe) [reply sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala]: Ok, removed file formats other than csv

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala [hunk 6]
```diff
@@ -214,4 +216,6 @@ class MockFileSystem extends RawLocalFileSystem {
   override def globStatus(pathPattern: Path): Array[FileStatus] = {
     mockGlobResults.getOrElse(pathPattern, Array())
   }
+
+  override def getUri: URI = URI.create("mockFs://mockFs/")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:220]: is this change needed?

File: sql/core/src/test/resources/test-data/test-archive.har/_index [hunk 7]
```diff
@@ -0,0 +1,7 @@
+%2F dir 1697722622766+493+tigrulya+hadoop 0 0 test.txt test.orc test.parquet test.json test.csv
```
Comment (cloud-fan) [inline sql/core/src/test/resources/test-data/test-archive.har/_index:1]: can we clean up the test files a bit? We only test csv now.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala (hunk 6 above)
Comment (tigrulya-exe) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceSuite.scala:220]: Yes, if we don't override this method, then path check inside `fs.makeQualified(path)` will fail, because it expects path with `file://` scheme (`MockFileSystem` inherits `RawLocalFileSystem`)

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!
